lib/cli/src/commands/publish.rs

```diff
@@ -0,0 +1,652 @@
+        manifest.base_directory_path = cwd.clone();
+
+        if let Some(package_name) = self.package_name.as_ref() {
+            manifest.package.name = package_name.to_string();
+        }
+
+        if let Some(version) = self.version.as_ref() {
+            manifest.package.version = version.clone();
+        }
+
+        let registry = match self.registry.as_deref() {
+            Some(s) => wasmer_registry::format_graphql(s),
+            None => {
+                let config = PartialWapmConfig::from_file()
+                    .map_err(|e| anyhow::anyhow!("could not load config {e}"))?;
+                config.registry.get_current_registry()
+            }
+        };
+
+        if !self.no_validate {
+            validate::validate_directory(&manifest, &registry, cwd.clone())?;
+        }
+
+        builder.append_path_with_name(&manifest_path_buf, PACKAGE_TOML_FALLBACK_NAME)?;
+
+        let manifest_string = toml::to_string(&manifest)?;
+
+        let (readme, license) = construct_tar_gz(&mut builder, &manifest, &cwd)?;
+
+        builder.finish().ok();
+        let tar_archive_data = builder.into_inner().map_err(|_| PublishError::NoModule)?;
+        let archive_name = "package.tar.gz".to_string();
+        let archive_dir = tempfile::TempDir::new()?;
+        let archive_dir_path: &std::path::Path = archive_dir.as_ref();
+        fs::create_dir(archive_dir_path.join("wapm_package"))?;
+        let archive_path = archive_dir_path.join("wapm_package").join(&archive_name);
+        let mut compressed_archive = fs::File::create(&archive_path).unwrap();
+        let mut gz_enc = GzEncoder::new(&mut compressed_archive, Compression::best());
+
+        gz_enc.write_all(&tar_archive_data).unwrap();
+        let _compressed_archive = gz_enc.finish().unwrap();
+        let mut compressed_archive_reader = fs::File::open(&archive_path)?;
+
+        let maybe_signature_data = sign_compressed_archive(&mut compressed_archive_reader)?;
+        let archived_data_size = archive_path.metadata()?.len();
+
+        assert!(archive_path.exists());
+        assert!(archive_path.is_file());
+
+        if self.dry_run {
+            // dry run: publish is done here
+
+            println!(
+                "Successfully published package `{}@{}`",
+                manifest.package.name, manifest.package.version
+            );
+
+            log::info!(
+                "Publish succeeded, but package was not published because it was run in dry-run mode"
+            );
+
+            return Ok(());
+        }
+
+        wasmer_registry::publish::try_chunked_uploading(
+            Some(registry),
+            self.token.clone(),
+            &manifest.package,
+            &manifest_string,
+            &license,
+            &readme,
+            &archive_name,
+            &archive_path,
+            &maybe_signature_data,
+            archived_data_size,
+            self.quiet,
+        )
+        .map_err(on_error)
```

Comment:
Is it worth hoisting the `on_error()` cal into the main `cli` module so we only need to report errors to sentry once, rather than  needing to do it for every sub-command?

Reply:
We could although I'm not sure we'd want to report every failure of every subcommand to sentry. I think this is fine, we just want to see if the publishing fails (likely because of wapm having an error). I don't think we want to log every failure of every subcommand, in fact this is the only subcommand where this logging is implemented.

Reply:
Sounds reasonable to me. I imagine we'd get spammed quite a lot if we were reporting to sentry whenever a package lookup fails or we can't run a `*.wasm` file for whatever reason.